### PR TITLE
MainTableView+NetworkOperations

### PR DESCRIPTION
**Pull request by Alexdive**

### CODE REVIEW

File: PokemonList/MainTableVC/MainTableVCViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+//  Created by Alex Permiakov on 4/19/21.
+//
+
+import UIKit
+
+protocol MainTableVCViewModelDelegate: class {
+    func onFetchCompleted(with newIndexPathsToReload: [IndexPath]?)
+}
+
+final class MainTableVCViewModel {
```
Comment (kliukin) [inline PokemonList/MainTableVC/MainTableVCViewModel.swift:14]: This VM should be closed with a protocol.

File: PokemonList/MainTableVC/MainTableVCViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                    } catch let error {
+                        print(error.localizedDescription)
+                    }
+                }
+            }
+            queue.addOperation(downloadOp)
+        }
+    }
+   
+    func getPokemon(with indexPath: IndexPath, completion: @escaping ((Pokemon) -> Void)) {
```
Comment (kliukin) [inline PokemonList/MainTableVC/MainTableVCViewModel.swift:74]: Let's break down this method as well.

File: PokemonList/MainTableVC/MainTableViewController.swift [hunk 3]
```diff
@@ -0,0 +1,151 @@
+//
+//  MainTableViewController.swift
+//  PokemonList
+//
+//  Created by Alex Permiakov on 4/16/21.
+//
+
+import UIKit
+
+final class MainTableViewController: UIViewController {
```
Comment (kliukin) [inline PokemonList/MainTableVC/MainTableViewController.swift:10]: This view should be closed with a protocol.

File: PokemonList/MainTableVC/MainTableVCViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,149 @@
+//
+//  MainTableVCViewModel.swift
+//  PokemonList
+//
+//  Created by Alex Permiakov on 4/19/21.
+//
+
+import UIKit
```
Comment (kliukin) [inline PokemonList/MainTableVC/MainTableVCViewModel.swift:8]: Normally you don't want to use UIKit classes in VMs.

File: PokemonList/MainTableVC/MainTableVCViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    
+    func getList(with stringUrl: String) {
+        
+        guard !isFetchInProgress else {
+            return
+        }
+        isFetchInProgress = true
+        
+        if let url = URL(string: stringUrl) {
+            let downloadOp = NetworkDataOperation(url: url) { [weak self] result in
```
[the remark this answers is not in the document]
Comment (kliukin) [reply PokemonList/MainTableVC/MainTableVCViewModel.swift:39]: Also your VM should be injected with dependencies via init. 
Something like:
`init(cacher: CacherProtocol, networkManager: NetworkManagerProtocol)`

File: PokemonList/MainTableVC/MainTableViewController.swift [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    var viewModel: MainTableVCViewModel
+    
+    let startUrlString = "https://pokeapi.co/api/v2/pokemon"
```
Comment (kliukin) [inline PokemonList/MainTableVC/MainTableViewController.swift:14]: The view should not have this data.

The responsibilities of you views are:
1) Show UI for data from VM
2) Handle users interactions

**Review by kliukin — CHANGES_REQUESTED** [review]
1. Let's try to make smaller PRs.
2. Read some articles about MVVM approach (like https://www.toptal.com/ios/swift-tutorial-introduction-to-mvvm) and refine your solution. 
3. Since it's MVVM let's add some reactive binding. Maybe with Combine if it's possible.

File: PokemonList/Cache.swift [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+class StructWrapper<T>: NSObject {
+
+    let value: T
+
+    init(convertStruct: T) {
+        self.value = convertStruct
+    }
+}
+
+struct Cache {
```
[the remark this answers is not in the document]
Comment (Alexdive) [reply PokemonList/Cache.swift:19]: No idea :(

File: PokemonList/MainTableVC/MainTableVCViewModel.swift (hunk 4 above)
Comment (Alexdive) [reply PokemonList/MainTableVC/MainTableVCViewModel.swift:8]: How to return downloaded image then?

**Comment by Alexdive** [comment]
Combine won't work if our target is ios11

**Pull request merged**